### PR TITLE
MM-22800 - Show Keyboard Shortcuts on Tooltips

### PR DESCRIPTION
**Pull request by asaadmahmood**
#### Summary
MM-22800 - Show Keyboard Shortcuts on Tooltips

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/15024

#### Screenshots
Screenshot 2021-05-26 at 11 04 53 AM

```release-note
NONE
```

### CODE REVIEW

**Comment by asaadmahmood** [comment]
@esethna There were a lot of conflicts to work with, had to understand what was happening and consolidate some of the new stuff, so opened up a new PR for this.

**Comment by esethna** [comment]
Thanks Asaad! Adding @matthewbirtch for UX review.

Question: is it intentional that we are changing the tooltip text styling? 

image
image

**Review by esethna — APPROVED** [review]
Approved from PM perspective

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
@asaadmahmood can we adjust the `font-size` on the actual shortcuts to be 12px? They are larger than the tooltip text. The font size should be 12px for everything within the tooltip.

I'm curious about the question from @esethna as well. If we're updating typography here, we could take the opportunity to align better with the Figma component. However, UI platform team will eventually be taking this on, so I don't want to double up efforts.

Other than that, I think we're good.

**Comment by asaadmahmood** [comment]
@matthewbirtch @esethna I was just reusing the styles that were there on the other PR, but I can revert those back. It's using 600 font weight, similar to what we have in the FIgma esigns.

Screenshot 2021-05-26 at 11 40 26 PM

**Comment by asaadmahmood** [comment]
@matthewbirtch I've reduced the font size for the shortcuts, if we want to change the font weight from 600 to 400 again for the text, let me know.

**Comment by matthewbirtch** [comment]
> @matthewbirtch @esethna I was just reusing the styles that were there on the other PR, but I can revert those back. It's using 600 font weight, similar to what we have in the FIgma esigns.
> 
> Screenshot 2021-05-26 at 11 40 26 PM

That's fine. 600 font-weight is fine, but should we update the font-family to open sans to match the latest component designs while we're at it?

**Comment by asaadmahmood** [comment]
@matthewbirtch Done.

**Review by matthewbirtch — APPROVED** [review]
Looks great @asaadmahmood. Thanks!

**Comment by esethna** [comment]
Thanks Asaad! Adding UI team for reviews. @michelengelen @nevyangelova

**Comment by esethna** [comment]
@amyblais regarding release note label

**Comment by amyblais** [comment]
@esethna PR owner owns adding release notes. cc @asaadmahmood

File: components/shortcuts/shortcuts_modal/shortcuts_modal.test.tsx
Comment (nevyangelova) [inline components/shortcuts/shortcuts_modal/shortcuts_modal.test.tsx:13]: we should probably test if the shortcuts are rendered correctly too

**Review by nevyangelova — CHANGES_REQUESTED** [review]

**Comment by esethna** [comment]
@nevyangelova what are the extend of the changes you've requested here and which of them would block the PR from going in? @asaadmahmood mentioned that he may need help to implement some of the suggestions. So wondering if there's a minimum set of changes that would be required to get this PR merged

File: components/shortcuts/shortcuts_modal/shortcuts_modal.test.tsx
Comment (asaadmahmood) [reply components/shortcuts/shortcuts_modal/shortcuts_modal.test.tsx:13]: Sounds cool, but I'm not that experienced with writing tests, so someone else may have to look into this.

File: components/shortcuts/shortcuts_modal/shortcuts_modal.tsx
Comment (asaadmahmood) [reply components/shortcuts/shortcuts_modal/shortcuts_modal.tsx:129]: We can do that, but the items in the array are kept without reference to the subcategory in the modal, so not sure how it can be done without rewriting the shortcuts to contain that information.

**Comment by michelengelen** [comment]
Sorry to butt in here: Changes requested from @nevyangelova are not that severe. This could be done in a timely manner. IMO these small changes should not be postponed since (from my experience) it is better to do it while it is still in development.

**Comment by asaadmahmood** [comment]
@michelengelen Completely agreed. Someone else may have to take these changes because some of the changes like the test may not be suitable for me, and the other one mapping the shortcuts record may be quite a larger change as that may require adding subsection information (from what I've understood) in the shortcuts record by default.

That shortcuts record change is something that's already implemented the way it is on master. This PR was about showing the keyboards shortcut on tooltip, but looks like we want to make a larger change that would change how the shortcut modal previously worked and modify the record as well.

So might as well have someone from the webapp team look at the whole thing if we're trying to improve how the modal itself is working.

As improving how the records are shown in the shortcuts modal wasn't really part of this ticket.

Below image is of master.
Screenshot 2021-06-08 at 11 32 10 PM

**Comment by nevyangelova** [comment]
@esethna I understand this might need to make the next release but the dev reviews are a necessary step for this process and I see you requested the review from @michelengelen as well so more change requests should be anticipated. I have kept the comments to a minimum in this case. @asaadmahmood we have several component test which are rendered with different props passed to the component in webapp feel free to use those as a reference. It might be a good learning opportunity in this case.

File: components/shortcuts/shortcuts_modal/shortcuts_modal.scss
Comment (michelengelen) [inline components/shortcuts/shortcuts_modal/shortcuts_modal.scss:69]: there is no need for nesting here

File: components/shortcuts/shortcuts_modal/shortcuts_modal.tsx
Comment (michelengelen) [inline components/shortcuts/shortcuts_modal/shortcuts_modal.tsx:57]: why is this built using a class-based component? Especially since this is a new component!

File: components/shortcuts/shortcuts_modal/shortcuts_modal.tsx
Comment (michelengelen) [inline components/shortcuts/shortcuts_modal/shortcuts_modal.tsx:85]:
```suggestion
    //toggles the state of shortcut dialog
    handleToggle = () => this.setState({
        show: !this.state.show,
    });
```

File: components/shortcuts/shortcuts_modal/shortcuts_modal.tsx
Comment (michelengelen) [inline components/shortcuts/shortcuts_modal/shortcuts_modal.tsx:89]:
```suggestion
    handleHide = () => this.setState({show: false});
```

File: components/sidebar/channel_navigator/channel_navigator.tsx
Comment (michelengelen) [inline components/sidebar/channel_navigator/channel_navigator.tsx:35]: why this change?

File: components/sidebar/channel_navigator/channel_navigator.tsx
Comment (michelengelen) [inline components/sidebar/channel_navigator/channel_navigator.tsx:119]:
```suggestion
        const getTooltip = (shortcut) => (
            <Tooltip
                id='upload-tooltip'
            >
                <ShortcutSequence
                    shortcut={shortcut}
                    hoistDescription={true}
                />
            </Tooltip>
        );
```

File: components/sidebar/channel_navigator/channel_navigator.tsx
Comment (michelengelen) [inline components/sidebar/channel_navigator/channel_navigator.tsx:128]:
```suggestion
                        overlay={getTooltip(shortcuts.browserChannelPrev)}
```

File: components/sidebar/channel_navigator/channel_navigator.tsx
Comment (michelengelen) [inline components/sidebar/channel_navigator/channel_navigator.tsx:143]:
```suggestion
                        overlay={getTooltip(shortcuts.browserChannelNext)}
```

File: i18n/en.json
Comment (michelengelen) [inline i18n/en.json:4237]: OS detection could help here as well.

**Review by michelengelen — CHANGES_REQUESTED** [review]
sry for all the comments, but IMO the component and data-structure is heavily over-complicated. IMO this is prone to errors. 

I would like to have one more opinion on this, so tagging @hmhealey for another review.

Sorry @esethna, but this seems far from ready to me!

File: components/channel_header/components/header_icon_wrapper.tsx
Comment (asaadmahmood) [reply components/channel_header/components/header_icon_wrapper.tsx:160]: Not sure @michelengelen. Most of the code is primarily coming from this PR:
https://github.com/mattermost/mattermost-webapp/pull/6138

It's a community PR, and I was just fixing a UI bug and conflicts on it so it could get in. Looks like the whole PR needs quite some work before it can be done.

**Comment by asaadmahmood** [comment]
Hey @esethna I agree with @michelengelen and @nevyangelova. This PR may need a bunch of work as the community PR was a bit far from done, so I'll be taking away my ownership from this PR, and someone from the core team can work on this, as there are things that perhaps the original contributor did, which I myself am not sure/experienced with.

**Comment by esethna** [comment]
> Sorry @esethna, but this seems far from ready to me!

That's fine, that what dev reviews are for. If we're not comfortable taking this and it's a large amount of work to fix it, let's close this PR and reopen the ticket

**Comment by michelengelen** [comment]
@esethna should we keep this PR open for future reference, or can we close it?

**Pull request closed without merging**
